fine-tune-lora.py

```
@@ -0,0 +1,142 @@
+import os
+from typing import Dict
+
+import torch
+from datasets import load_dataset
+from transformers import (
+    AutoTokenizer,
+    AutoModelForCausalLM,
+    TrainingArguments,
+    Trainer,
+)
+from peft import LoraConfig, get_peft_model
+import dotenv
+from torch.nn.utils.rnn import pad_sequence
+
+dotenv.load_dotenv()
+
+if torch.cuda.is_available():
+    print("✅ GPU detected:", torch.cuda.get_device_name(0))
+else:
+    print("⚠️ GPU not available, using CPU")
+
+
+# ----------  Prompt template ----------
+def make_example(example: Dict[str, str], tokenizer):
+    # build prompt and reply strings
+    prompt = f"<s>[EMAIL THREAD]\n{example['input']}\n\n[YOUR REPLY]\n"
+    reply = example["output"] + tokenizer.eos_token
+
+    # tokenize
+    prompt_ids = tokenizer(prompt, add_special_tokens=False).input_ids
+    reply_ids = tokenizer(reply, add_special_tokens=False).input_ids
+
+    # concat and build labels
+    input_ids = prompt_ids + reply_ids
+    labels = [-100] * len(prompt_ids) + reply_ids
+
+    return {"input_ids": input_ids, "labels": labels}
+
+
+# ----------  Main ----------
+def main():
+    # Tokeniser
+    tokenizer = AutoTokenizer.from_pretrained(
+        os.getenv("MODEL_NAME"),
+        trust_remote_code=True,
+    )
+    tokenizer.pad_token = tokenizer.eos_token
+    tokenizer.padding_side = "right"
+
+    model = AutoModelForCausalLM.from_pretrained(
+        os.getenv("MODEL_NAME"),
+        torch_dtype=torch.float16 if torch.cuda.is_available() else torch.float32,
+        trust_remote_code=True,
+        low_cpu_mem_usage=True,
+    )
+
+    if torch.cuda.is_available():
+        model.to("cuda")
```

Would fine tuning work at all without `cuda` given our resource requirements? We could just fail if its not available.

---

Yea.. we could. Just that sometimes its useful to run the script on CPU for debugging purposes as well.